arch/x86/core/intel64/irq.c

```diff
@@ -92,15 +92,11 @@ int z_arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
 
 void irq_offload(irq_offload_routine_t routine, void *parameter)
 {
-	u32_t key;
-
-	key = irq_lock();
```

Comment:
IIRC, the intent here was to serialize setting x86_irq_funcs/args so that concurrent entry to this function didn't clobber stuff. maybe a spinlock here instead?

this API is only used for tests, seems very unlikely in practice though.

Reply:
Well, the proper way to synchronize here has never been settled.  I don't think a spinlock is a good solution, because holding a spinlock should imply that interrupts are disabled on the local core (@andyross can confirm if this is the design intent).  We shouldn't *simulate* an interrupt under those circumstances, either.

On UP systems, the `irq_lock()`/`irq_unlock()` pairing is misleading -- interrupts are NOT disabled in between them, they're only disabled long enough to set the offload function and its argument, and then are re-enabled in the IRQ handler path, only to be briefly disabled again in the IRQ exit path, basically right before `irq_unlock()` is called to re-enable them.